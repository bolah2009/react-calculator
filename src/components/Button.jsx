import React from 'react';
import PropTypes from 'prop-types';
import styles from '../css/modules/Button.module.css';

const generateClassName = (buttonIsWide, stylesObject) => {
  const { button, wide } = stylesObject;
  if (buttonIsWide) {
    return `${button} ${wide}`;
  }
  return button;
};

const Button = ({ name, isWide, color, handleClick }) => (
  <button
    style={{ backgroundColor: color }}
    className={generateClassName(isWide, styles)}
    type="button"
    id={`button-${name}`}
    onClick={handleClick}
  >
    {name}
  </button>
);

Button.propTypes = {
  name: PropTypes.string.isRequired,
  isWide: PropTypes.bool.isRequired,
  handleClick: PropTypes.func.isRequired,
  color: PropTypes.string,
};

Button.defaultProps = {
  color: 'orange',
};

export default Button;
