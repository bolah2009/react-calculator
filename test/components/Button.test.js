import React from 'react';
import renderer from 'react-test-renderer';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import Button from '../../src/components/Button';

let container = null;
const handleClick = jest.fn();
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('Button component', () => {
  test('should render with orange default background color', () => {
    const button = renderer.create(
      <Button onClick={i => handleClick(i)} name="0" isWide={false} />,
    );
    expect(button.toJSON()).toMatchSnapshot();
  });

  test('should render with grey background color', () => {
    const button = renderer.create(
      <Button onClick={i => handleClick(i)} color="grey" name="+" isWide={false} />,
    );
    expect(button.toJSON()).toMatchSnapshot();
  });

  test('should render with 50% width when isWide is true', () => {
    const button = renderer.create(
      <Button onClick={i => handleClick(i)} color="grey" name="0" isWide />,
    );
    expect(button.toJSON()).toMatchSnapshot();
  });
});

it('renders with name and isWide property', () => {
  act(() => {
    render(<Button onClick={i => handleClick(i)} name="=" isWide />, container);
  });

  expect(container.textContent).toBe('=');
});

it('renders with or without color property', () => {
  act(() => {
    render(<Button onClick={i => handleClick(i)} name="=" isWide />, container);
  });
  expect(container.firstChild.style.backgroundColor).toBe('orange');

  act(() => {
    render(<Button onClick={i => handleClick(i)} color="lightgrey" name="=" isWide />, container);
  });

  expect(container.firstChild.style.backgroundColor).toBe('lightgrey');
});
