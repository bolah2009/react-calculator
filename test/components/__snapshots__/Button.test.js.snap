// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render with 50% width when isWide is true 1`] = `
<button
  className="button wide"
  id="button-0"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
  type="button"
>
  0
</button>
`;

exports[`Button component should render with grey background color 1`] = `
<button
  className="button"
  id="button-+"
  style={
    Object {
      "backgroundColor": "grey",
    }
  }
  type="button"
>
  +
</button>
`;

exports[`Button component should render with orange default background color 1`] = `
<button
  className="button"
  id="button-0"
  style={
    Object {
      "backgroundColor": "orange",
    }
  }
  type="button"
>
  0
</button>
`;
